File: leetcode/top-k-frequent-elements.java
// https://leetcode.com/problems/top-k-frequent-elements/

import java.util.HashMap;
import java.util.PriorityQueue;

class Solution {
    class Pair{
        int freq;
        int val;
        Pair(int f, int v){
            this.freq = f;
            this.val = v;
        }
    }
    
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> freq = new HashMap<Integer, Integer>();
        
        for(int i = 0; i<nums.length; i++){
            Integer count = freq.get(nums[i]);
            freq.put(nums[i], count == null ? 1 : count+1);
        }
        
        PriorityQueue<Pair> queue = new PriorityQueue<Pair>((a,b)->b.freq-a.freq);
        
        for(int i: freq.keySet()){
            queue.add(new Pair(freq.get(i), i));
        }
        
        int[] out = new int[k];
        
        for(int i = 0; i<k; i++){
            out[i] = queue.poll().val;
        }
        
        
        return out;
    }
}

File: leetcode/top_k_frequent_elements.py
# https://leetcode.com/problems/top-k-frequent-elements/


class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        freq: dict[int] = {}
        for i in nums:
            freq[i] = (i, freq.get(i)[1] + 1 if freq.get(i) is not None else 1)

        tuples = list(freq.values())
        tuples.sort(key=lambda a: a[1], reverse=True)

        out = []
        for i in tuples:
            out.append(i[0])
            if len(out) == k:
                return out


class SolutionTwo:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        freq_map: dict[int, int] = {}
        for i in nums:
            freq_map[i] = freq_map.get(i, 0) + 1

        frequencies: list[list[int]] = sorted(list(freq_map.items()), key=lambda x: -x[1])

        results: list[int] = []
        for i in range(k):
            results.append(frequencies[i][0])

        return results
